src/components/structures/ScrollPanel.js

```diff
@@ -128,13 +153,76 @@ module.exports = React.createClass({
     checkFillState: function() {
         var sn = this._getScrollNode();
 
+        // if there is less than a screenful of messages above or below the
+        // viewport, try to get some more messages.
+        //
+        // scrollTop is the number of pixels between the top of the content and
+        //     the top of the viewport.
+        //
+        // scrollHeight is the total height of the content.
+        //
+        // clientHeight is the height of the viewport (excluding borders,
+        // margins, and scrollbars).
+        //
+        //
+        //   .---------.          -                 -
+        //   |         |          |  scrollTop      |
+        // .-+---------+-.    -   -                 |
+        // | |         | |    |                     |
+        // | |         | |    |  clientHeight       | scrollHeight
+        // | |         | |    |                     |
+        // `-+---------+-'    -                     |
+        //   |         |                            |
+        //   |         |                            |
+        //   `---------'                            -
+        //
+
         if (sn.scrollTop < sn.clientHeight) {
-            // there's less than a screenful of messages left - try to get some
-            // more messages.
-            this.props.onFillRequest(true);
+            // need to back-fill
+            this._maybeFill(true);
+        }
+        if (sn.scrollTop > sn.scrollHeight - sn.clientHeight * 2) {
```

Comment:
I love the ASCII art, thanks! I still don't understand how `sn.clientHeight * 2` is meant to be used to determine if `// need to forward-fill` though. How does that work? I'd have thought you'd want to get the bottom height (the only thing not marked on the diagram, lines `:175-176`; `scrollHeight - clientHeight - scrollTop`) and then do a similar comparison to `:180` to see if that height is smaller than the client's viewport, so it would be:

```javascript
if ( (sn.scrollHeight - sn.clientHeight - sn.scrollTop) < sn.clientHeight) {
```

but that isn't what we're doing. Have I misunderstood the logic somewhere?